Add globalnet diagnose to subctl

yboaron opened the pull request:
Enhance the subctl diagnose code to verify if
Globalnet objects are properly created and matching
with the expected configuration.

Signed-off-by: yboaron <yboaron@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

yboaron committed:
Add globalnet diagnose to subctl

Enhance the subctl diagnose code to verify if
Globalnet objects are properly created and matching
with the expected configuration.

Signed-off-by: yboaron <yboaron@redhat.com>

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+		status.EndWithSuccess("Globalnet is enabled and properly configured")
+	} else {
+		status.EndWithSuccess("Globalnet is disabled")
+	}
```
I think we should check for enabled above. Also we can simplify the code by using the new `reporter.Interface` methods.

```suggestion
	if cluster.Submariner == nil {
		status.Warning(cmd.SubmMissingMessage)
		return true
	}
	
	if cluster.Submariner.Spec.GlobalCIDR == "" {
	    status.Success("Globalnet is not installed - skipping")
	    return true
	}

	status.Start("Checking Globalnet configuration")
	defer status.End()

	return checkClusterGlobalEgressIps(cluster, status) &&
		   checkGlobalEgressIps(cluster, status) &&
		   checkGlobalIngressIps(cluster, status)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+	return true
+}
+
+func checkClusterGlobalEgressIps(cluster *cmd.Cluster, status *cli.Status) {
```
```suggestion
func checkClusterGlobalEgressIps(cluster *cmd.Cluster, status reporter.Interface) bool {
```

sridhargaddam replied:
+1, @yboaron you can create a reporter interface for stdout
```
status := reporter.Stdout()
```

tpantelis replied:
We want the caller to use a CLI , just pass in as a `reporter.Interface`

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+		corev1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		status.QueueFailureMessage(fmt.Sprintf("Error obtaining the clusterGlobalEgressIPs cr: %v", err))
+		return
```
```suggestion
		status.Failure("Error listing the ClusterGlobalEgressIP resources: %v", err)
		return false
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	if len(clusterGlobalEgress.Items) != 1 {
+		status.QueueFailureMessage(fmt.Sprintf("Number (%d) of clusterGlobalEgress resources != 1", len(clusterGlobalEgress.Items)))
+		return
```
I think this can be a warning since we only process the default instance. Also don't return here.

```suggestion
		status.Warning("Found %d ClusterGlobalEgressIP resources but only the default instance is supported", len(clusterGlobalEgress.Items))
```

sridhargaddam replied:
We can also print the name of the default instance (i.e., "cluster-egress.submariner.io") to make it easy for the user.

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	if clusterGlobalEgress.Items[0].Name != constants.ClusterGlobalEgressIPName {
+		status.QueueFailureMessage(fmt.Sprintf("clusterGlobalEgress name (%s) not equals to default name", clusterGlobalEgress.Items[0].Name))
+		return
```
```suggestion
		status.Failure("The name of the ClusterGlobalEgressIP resource (%s) does not match the expected default name (%s)", clusterGlobalEgress.Items[0].Name, constants.ClusterGlobalEgressIPName)
		failed := true
```

sridhargaddam replied:
What if the user does have the default clusterGlobalEgressIP object but has also created an additional one accidentally. Here we are checking if the zeroth entry `clusterGlobalEgress.Items[0].Name` is matching the default one. We should modify the check.

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+	if clusterGlobalEgress.Items[0].Name != constants.ClusterGlobalEgressIPName {
+		status.QueueFailureMessage(fmt.Sprintf("clusterGlobalEgress name (%s) not equals to default name", clusterGlobalEgress.Items[0].Name))
+		return
+	}
```
I think we should check if the number of IPs requested equals the number allocated. Also verify that the status condition is success.

```
import (
    "k8s.io/apimachinery/pkg/api/meta"
    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)
```

```
clusterGlobalEgressIP := clusterGlobalEgress.Items[0]
...

numberOfIPs := 1
if clusterGlobalEgressIP.Spec.NumberOfIPs != nil {
    numberOfIPs = *clusterGlobalEgressIP.Spec.NumberOfIPs
}

if numberOfIPs != len(clusterGlobalEgressIP.Status.AllocatedIPs) { 
    status.Failure("The number of requested IPs (%d) does not match the number allocated (%d) for ClusterGlobalEgressIP %q", 
        numberOfIPs, len(clusterGlobalEgressIP.Status.AllocatedIPs), clusterGlobalEgressIP.Name)
    failed := true
}

condition := meta.FindStatusCondition(clusterGlobalEgressIP.Status.Conditions, submarinerv1.GlobalEgressIPAllocated)
if condition == nil {
    status.Failure("ClusterGlobalEgressIP %q is missing the %q status condition %q", clusterGlobalEgressIP.Name,
        submarinerv1.GlobalEgressIPAllocated)
    failed := true
}

if condition.Status != metav1.ConditionTrue {
    status.Failure("The allocation of global IPs for ClusterGlobalEgressIP %q failed with reason %q and message %q", 
        clusterGlobalEgressIP.Name, condition.Reason, condition.Message)
    failed := true
}

return failed
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+	}
+}
+
+func checkGlobalEgressIps(cluster *cmd.Cluster, status *cli.Status) {
```
```suggestion
func checkGlobalEgressIps(cluster *cmd.Cluster, status reporter.Interface) bool
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+		if *globalEgressIps.Items[i].Spec.NumberOfIPs != len(globalEgressIps.Items[i].Status.AllocatedIPs) {
+			status.QueueFailureMessage(fmt.Sprintf("Error GlobalEgressIPs(%s) NumberOfIPs != AllocatedIPs", globalEgressIps.Items[i].Name))
+			return
+		}
```
The `NumberOfIPs` could be nil. Also check the status condition the same as I outlined above for the `ClusterGlobalEgressIP`.

```suggestion
                gip := globalEgressIps.Items[i]
                numberOfIPs := 1
                if gip.Spec.NumberOfIPs != nil {
                    numberOfIPs = *gip.Spec.NumberOfIPs
                }

		if numberOfIPs != len(gip.Status.AllocatedIPs) {
			status.Failure("The number of requested IPs (%d) does not match the number allocated (%d) for GlobalEgressIP %q",
                             numberOfIPs, len(gip.Status.AllocatedIPs), gip.Name)
		}
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+	})
+}
+
+func checkGlobalNet(cluster *cmd.Cluster) bool {
```
```suggestion
func checkGlobalnet(cluster *cmd.Cluster) bool {
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+	}
+}
+
+func checkGlobalIngressIps(cluster *cmd.Cluster, status *cli.Status) {
```
```suggestion
func checkGlobalIngressIps(cluster *cmd.Cluster, reporter.Interface) bool
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+	serviceExports, err := mcsClient.MulticlusterV1alpha1().ServiceExports(corev1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		status.QueueFailureMessage(fmt.Sprintf("Error obtaining ServiceExport resources: %v", err))
+		return
```
```suggestion
		status.Failure("Error listing ServiceExport resources: %v", err)
		return false
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+		if err != nil {
+			status.QueueFailureMessage(fmt.Sprintf("Error matching Service resource for exported service(%s,%s).%v", ns, name, err))
+			return
+		}
```
The `ServiceExport` can exist without the `Service` so emit a warning.

**Note**: `import apierrors "k8s.io/apimachinery/pkg/api/errors"`

```suggestion
		if apierrors.IsNotFound(err)  {
		        status.Warning("No matching Service resource found for exported service \"%s/%s\"", ns, name)
			continue
		}
		
		if err != nil {
		    status.Failure("Error retrieving Service: %v", err)
		    return false
		}
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+			return
+		}
+
+		if svc.Spec.Type == corev1.ServiceTypeClusterIP {
```
```suggestion
		if svc.Spec.Type != corev1.ServiceTypeClusterIP {
		    continue
		}
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+			if err != nil {
+				status.QueueFailureMessage(fmt.Sprintf("Error no GlobalIngressIP resource allocated for service(%s,%s).%v", ns, name, err))
+				return
+			}
```
```suggestion
                        if apierrors.IsNotFound(err)  {
		            status.Failure("No matching GlobalIngressIP resource found for exported service \"%s/%s\"", ns, name)
		            failed = true
			    continue
		        }
		
		        if err != nil {
		            status.Failure("Error retrieving GlobalIngressIP: %v", err)
		            return false
		        }
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+				status.QueueFailureMessage(fmt.Sprintf("Error no GlobalIngressIP resource allocated for service(%s,%s).%v", ns, name, err))
+				return
+			}
+
```
Check the status condition here as outlined above for `ClusterGlobalEgressIP`.

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+
+			if globalIngress.Status.AllocatedIP == "" {
+				status.QueueFailureMessage(fmt.Sprintf("Error GlobalIngressIP(%s,%s) - AllocatedIP is empty", ns, name))
+				return
```
`AllocatedIP` **shouldn't** be empty if the status condition indicates success but doesn't hurt to verify.

```suggestion
				status.Failure("No global IP was allocated for the GlobalIngressIP associated with exported service \"%s/%s\"", ns, name)
				failed = true
				continue
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+				context.TODO(), metav1.ListOptions{LabelSelector: fmt.Sprintf("submariner.io/exportedServiceRef=%s", name)})
+			if err != nil {
+				status.QueueFailureMessage(fmt.Sprintf("Error finding internal service for exported service(%s,%s).%v", ns, name, err))
+				return
```
```suggestion
				status.Failure("Error listing Services: %v", err)
		                 return false
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+			if len(svcs.Items) != 1 {
+				status.QueueFailureMessage(fmt.Sprintf("Error finding internal service for exported service(%s,%s).%v", ns, name, err))
+				return
+			}
```
```suggestion
			if len(svcs.Items) == 0 {
				status.Failure("No internal service found for exported service \"%s/%s\"", ns, name)
				failed = true
				continue
			}
			
			if len(svcs.Items) > 1 {
				status.Failure("Found %d internal services for exported service \"%s/%s\" - expected 1", len(svcs.Items), ns, name)
				failed = true
				continue
			}
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,162 @@
+			}
+
+			if svcs.Items[0].Spec.ExternalIPs[0] != globalIngress.Status.AllocatedIP {
+				status.QueueFailureMessage(fmt.Sprintf("GlobalIngressIP != ExternalIP of internal Service for exported service(%s,%s)", ns, name))
+				return
```
```suggestion
				status.Failure("The external IP (%s) for the internal service associated with exported service \"%s/%s\" does not match the allocated global IP (%s) stored in GlobalIngressIP %q", 
				    svcs.Items[0].Spec.ExternalIPs[0], ns, name, globalIngress.Status.AllocatedIP, globalIngress.Name)
				failed = true
```

sridhargaddam reviewed (changes requested):
Thanks for the PR @yboaron

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+	status.Start("Checking Globalnet configuration")
+	defer status.End()
+
+	retValue := checkClusterGlobalEgressIps(cluster, status) &&
+		checkGlobalEgressIps(cluster, status) &&
+		checkGlobalIngressIps(cluster, status)
+
+	if retValue {
+		status.EndWithSuccess("Globalnet is enabled and properly configured")
+	}
+
+	return retValue
```
With https://github.com/submariner-io/submariner-operator/pull/1876 we can do:

```suggestion
        tracker := reporter.NewTracker(status)
        
	checkClusterGlobalEgressIps(cluster, tracker)
        checkGlobalEgressIps(cluster, tracker)
	checkGlobalIngressIps(cluster, tracker)

	return !tracker.HasFailures()
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+	clusterGlobalEgress, err := cluster.SubmClient.SubmarinerV1().ClusterGlobalEgressIPs(
+		corev1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		status.Warning("Error listing the ClusterGlobalEgressIP resources: %v", err)
```
```suggestion
		status.Failure("Error listing the ClusterGlobalEgressIP resources: %v", err)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+	globalEgressIps, err := cluster.SubmClient.SubmarinerV1().GlobalEgressIPs(
+		corev1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		status.Warning("Error obtaining GlobalEgressIPs resources: %v", err)
```
```suggestion
		status.Failure("Error obtaining GlobalEgressIPs resources: %v", err)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+
+	clusterGlobalEgressIP := clusterGlobalEgress.Items[0]
+
+	numberOfIPs := -1
```
The default is 1.

```suggestion
	numberOfIPs := 1
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+
+	for i := range globalEgressIps.Items {
+		gip := globalEgressIps.Items[i]
+		numberOfIPs := -1
```
```suggestion
		numberOfIPs := 1
```

yboaron replied:
Any special reason to set 1? 
I think -1 is a better option
since -1 is always != len(gip.Status.AllocatedIPs) 
I mean in case 'gip.Spec.NumberOfIPs' is nill

tpantelis replied:
We're defaulting it to 1. We don't require the user to set `gip.Spec.NumberOfIPs` hence it's a pointer and thus can be nil if the user didn't set it, in which case we default to 1.

yboaron replied:
Ack

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+				numberOfIPs, len(gip.Status.AllocatedIPs), gip.Name)
+
+			retValue = false
+		}
```
We should check the status condition as on L119 above.

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+func checkGlobalIngressIps(cluster *cmd.Cluster, status reporter.Interface) bool {
+	mcsClient, err := mcsClientset.NewForConfig(cluster.Config)
+	if err != nil {
+		status.Warning("Error obtaining mcs client: %v", err)
```
```suggestion
		status.Failure("Error obtaining mcs client: %v", err)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+
+	serviceExports, err := mcsClient.MulticlusterV1alpha1().ServiceExports(corev1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		status.Warning("Error listing ServiceExport resources: %v", err)
```
```suggestion
		status.Failure("Error listing ServiceExport resources: %v", err)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+	return retValue
+}
+
+func checkClusterGlobalEgressIps(cluster *cmd.Cluster, status reporter.Interface) bool {
```
```suggestion
func checkClusterGlobalEgressIPs(cluster *cmd.Cluster, status reporter.Interface) {
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+
+		if svcs.Items[0].Spec.ExternalIPs[0] != globalIngress.Status.AllocatedIP {
+			status.Failure(
+				"The external IP (%s) for internal svc associated with exported svc \"%s/%s\" doesn't match allocated IP (%s) in GlobalIngressIP %q",
```
```suggestion
				"The external IP (%s) for internal svc associated with exported service \"%s/%s\" doesn't match allocated IP (%s) in GlobalIngressIP %q",
```

yboaron replied:
Yeah, I changed to 'svc' because I didn't find how to split long string into multiple lines (to avoid lint error)
Any idea how to tackle that?

tpantelis replied:
You can use + to split it, eg

```
"The external IP (%s) for internal svc associated with exported service \"%s/%s\" doesn't match " +
"allocated IP (%s) in GlobalIngressIP %q"

```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+	return retValue
+}
+
+func checkGlobalEgressIps(cluster *cmd.Cluster, status reporter.Interface) bool {
```
```suggestion
func checkGlobalEgressIPs(cluster *cmd.Cluster, status reporter.Interface) {
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+	return retValue
+}
+
+func checkGlobalIngressIps(cluster *cmd.Cluster, status reporter.Interface) bool {
```
```suggestion
func checkGlobalIngressIPs(cluster *cmd.Cluster, status reporter.Interface) {
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,254 @@
+			status.Failure("Error retrieving GlobalIngressIP for exported service \"%s/%s\": %v", ns, name, err)
+			return false
+		}
+
```
Check the status condition here.

yboaron replied:
Not sure that I understood this comment

tpantelis replied:
I mean the GlobalIngressIP status Condition as was done with the *EgressIPs (c. L119).  If it couldn't allocate the IP for some reason it will set the status Condition to indicate that.

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+	checkGlobalIngressIps(cluster, tracker)
+
+	if !tracker.HasFailures() {
+		status.EndWithSuccess("Globalnet is enabled and properly configured")
```
```suggestion
		status.Success("Globalnet is properly configured and functioning")
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+	return !tracker.HasFailures()
+}
+
+func checkClusterGlobalEgressIps(cluster *cmd.Cluster, status reporter.Interface) {
```
```suggestion
func checkClusterGlobalEgressIPs(cluster *cmd.Cluster, status reporter.Interface) {
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+
+	if len(clusterGlobalEgress.Items) != 1 {
+		status.Failure(
+			"Found %d ClusterGlobalEgressIP resources but only the default instance (%s) is supported",
```
```suggestion
			"Found %d ClusterGlobalEgressIP resources but only the default instance %q is supported",
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+	}
+
+	if !foundDefaultResource {
+		status.Failure("Couldn't find the default ClusterGlobalEgressIP resource(%s)", constants.ClusterGlobalEgressIPName)
```
```suggestion
		status.Failure("Couldn't find the default ClusterGlobalEgressIP resource %q)", constants.ClusterGlobalEgressIPName)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+	}
+}
+
+func checkGlobalEgressIps(cluster *cmd.Cluster, status reporter.Interface) {
```
```suggestion
func checkGlobalEgressIPs(cluster *cmd.Cluster, status reporter.Interface) {
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+
+		condition := meta.FindStatusCondition(gip.Status.Conditions, string(submarinerv1.GlobalEgressIPAllocated))
+		if condition == nil {
+			status.Failure("globalEgress %q is missing the %q status condition", gip.Name, submarinerv1.GlobalEgressIPAllocated)
```
```suggestion
			status.Failure("GlobalEgressIP %q is missing the %q status condition", gip.Name, submarinerv1.GlobalEgressIPAllocated)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+			status.Failure("globalEgress %q is missing the %q status condition", gip.Name, submarinerv1.GlobalEgressIPAllocated)
+			continue
+		} else if condition.Status != metav1.ConditionTrue {
+			status.Failure("The allocation of global IPs for globalEgress %q failed with reason %q and message %q",
```
```suggestion
			status.Failure("The allocation of global IPs for GlobalEgressIP %q failed with reason %q and message %q",
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+	}
+}
+
+func checkGlobalIngressIps(cluster *cmd.Cluster, status reporter.Interface) {
```
```suggestion
func checkGlobalIngressIPs(cluster *cmd.Cluster, status reporter.Interface) {
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+		condition := meta.FindStatusCondition(globalIngress.Status.Conditions, string(submarinerv1.GlobalEgressIPAllocated))
+		if condition == nil {
+			status.Failure("globalIngress %q is missing the %q status condition", globalIngress.Name,
+				submarinerv1.GlobalEgressIPAllocated)
```
```suggestion
			status.Failure("GlobalIngressIP %q associated with exported service \"%s/%s\" is missing the %q status condition", globalIngress.Name, ns, name,
				submarinerv1.GlobalEgressIPAllocated)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+			continue
+		} else if condition.Status != metav1.ConditionTrue {
+			status.Failure("The allocation of global IPs for globalIngress %q failed with reason %q and message %q",
+				globalIngress.Name, condition.Reason, condition.Message)
```
```suggestion
			status.Failure("The allocation of global IPs for GlobalIngressIP %q associated with exported service \"%s/%s\" failed with reason %q and message %q",
				globalIngress.Name, ns, name, condition.Reason, condition.Message)
```

tpantelis commented on pkg/subctl/cmd/diagnose/globalnet.go:
```diff
@@ -0,0 +1,251 @@
+
+	if svcs.Items[0].Spec.ExternalIPs[0] != globalIngress.Status.AllocatedIP {
+		status.Failure(
+			"The external IP (%s) for internal service associated with exported svc \"%s/%s\" doesn't"+
```
```suggestion
			"The external IP (%s) for internal service associated with exported service \"%s/%s\" doesn't"+
```

sridhargaddam reviewed (commented):
Thanks for addressing the review comments @yboaron. There is one small comment to be addressed, please take a look.

yboaron commented:
> Thanks for addressing the review comments @yboaron. There is one small comment to be addressed, please take a look.

Ooops, I missed this one.
Done, 10x!

sridhargaddam reviewed (approved):
Thanks @yboaron

nyechiel committed:
Merge branch 'devel' into gn_diagnose
